src/app/modules/form/admin-form/admin-form.service.ts

```diff
@@ -790,46 +791,70 @@ export const updateFormCollaborators = (
-    Promise.all(
-      updatedCollaboratorEmails.map(async (email) => {
-        const emailDomain = email.split('@').pop()
-        const result = await AgencyModel.findOne({ emailDomain })
-        return !!result
-      }),
+  // find removed collaborators (i.e. collaborators in the original list)
+  // but has been removed in the updated list
+  const removedCollaboratorEmails = form.permissionList.filter((collaborator) =>
+    updatedCollaborators.every(
+      (newCollaborator) => collaborator.email !== newCollaborator.email,
     ),
-    (error) => {
-      logger.error({
-        message: 'Error encountered while validating new form collaborators',
-        meta: logMeta,
-        error,
-      })
-      return transformMongoError(error)
-    },
   )
-    .andThen((doNewCollaboratorsExist) => {
-      const falseIdx = doNewCollaboratorsExist.findIndex((exists) => !exists)
-      return falseIdx < 0
-        ? okAsync(undefined)
-        : errAsync(
-            new InvalidCollaboratorError(updatedCollaboratorEmails[falseIdx]),
-          )
-    })
-    .andThen(() =>
-      ResultAsync.fromPromise(
-        form.updateFormCollaborators(updatedCollaborators),
-        (error) => {
-          logger.error({
-            message: 'Error encountered while updating form collaborators',
-            meta: logMeta,
-            error,
-          })
-          return transformMongoError(error)
-        },
+
+  return (
+    ResultAsync.fromPromise(
+      // Check that all updated collaborator domains exist in the Agency collection.
+      Promise.all(
+        updatedCollaboratorEmails.map(async (email) => {
+          const emailDomain = email.split('@').pop()
+          const result = await AgencyModel.findOne({ emailDomain })
+          return !!result
+        }),
       ),
+      (error) => {
+        logger.error({
+          message: 'Error encountered while validating new form collaborators',
+          meta: logMeta,
+          error,
+        })
+        return transformMongoError(error)
+      },
     )
-    .andThen(({ permissionList }) => okAsync(permissionList))
+      .andThen((doNewCollaboratorsExist) => {
+        const falseIdx = doNewCollaboratorsExist.findIndex((exists) => !exists)
+        return falseIdx < 0
+          ? okAsync(undefined)
+          : errAsync(
+              new InvalidCollaboratorError(updatedCollaboratorEmails[falseIdx]),
+            )
+      })
+      // after checking that collaborator exists
+      // remove forms from workspaces for collaborators that were removed
+      .andThen(() => {
+        removedCollaboratorEmails.map(async (collaborator) => {
+          await UserService.findUserByEmail(collaborator.email).map(
+            async (user) =>
+              await removeFormFromAllWorkspaces({
+                formId: form._id,
+                userId: user._id,
+              }),
+          )
+        })
+        return okAsync(undefined)
```

Comment:
Do we need to wrap this in await Promise.all? otherwise removedCollaboratorEmails.map returns Promise[], and we would return okAsync(undefined) immediately

Reply:
Yep thanks for catching this Shuli! Should definitely be more defensive! [8fbeaac](https://github.com/opengovsg/FormSG/pull/6592/commits/8fbeaac340f60b28ed48940165dc546d4a84d88c)